tchap_red_list/__init__.py

```diff
@@ -11,36 +11,382 @@
+    ):
         # Keep a reference to the config and Module API
         self._api = api
         self._config = config
 
+        # Register callbacks
+        self._api.register_account_data_callbacks(
+            on_account_data_updated=self.update_red_list_status,
+        )
+
+        self._api.register_spam_checker_callbacks(
+            check_username_for_spam=self.check_user_in_red_list,
+        )
+
+        if setup_db:
+            # Set up the storage layer
+            run_in_background(self._setup_db)
+
+        if self._config.use_email_account_validity:
+            self._api.looping_background_call(self._add_expired_users, 60 * 60 * 1000)
+            self._api.looping_background_call(
+                self._remove_renewed_users, 60 * 60 * 1000
+            )
+
     @staticmethod
     def parse_config(config: Dict[str, Any]) -> RedListManagerConfig:
-        # Parse the module's configuration here.
-        # If there is an issue with the configuration, raise a
-        # synapse.module_api.errors.ConfigError.
-        #
-        # Example:
-        #
-        #     some_option = config.get("some_option")
-        #     if some_option is None:
-        #          raise ConfigError("Missing option 'some_option'")
-        #      if not isinstance(some_option, str):
-        #          raise ConfigError("Config option 'some_option' must be a string")
-        #
-        return RedListManagerConfig()
+        return RedListManagerConfig(**config)
+
+    async def update_red_list_status(
+        self,
+        user_id: str,
+        room_id: Optional[str],
+        account_data_type: str,
+        content: JsonDict,
+    ) -> None:
+        """Update a user's status in the red list when their account data changes.
+        Implements the on_account_data_updated account data callback.
+        """
+        if account_data_type != ACCOUNT_DATA_TYPE:
+            return
+
+        # Compare what status (in the list, not in the list) the user wants to have with
+        # what it already has. If they're the same, don't do anything more.
+        desired_status = content.get("hide_profile") or False
+        current_status, because_expired = await self._get_user_status(user_id)
+
+        # If the user's desired status is the same as their current one, don't do
+        # anything, except if because_expired is True (because in this case it can be a
+        # user that was added to the red list after expiring, haven't been removed yet,
+        # and manually re-added themselves).
+        if current_status == desired_status and because_expired is False:
+            return
+
+        # Update the red list depending on whether the user wants their profile hidden.
+        if because_expired is True:
+            await self._update_added_after_expiring(user_id)
+        elif desired_status is True:
+            await self._add_to_red_list(user_id)
+        else:
+            await self._remove_from_red_list(user_id)
```

Comment:
If the user is in the red list because their account expired (`(current_status, because_expired) = (True, True)` and `desired_status` is `False`, ie. the user has for some reason reiterated that they do not want to be in the red list, then we'll call `_update_added_after_expiring` and incorrectly mark them as opting in to the red list.

Jiggling the ifs around would work:
```py
if current_status == desired_status:
    if because_expired:
        # Clear the because_expired flag
        self._update_added_after_expiring(user_id)
else:
    if desired_status is True:
        self._add_to_red_list(user_id)
    else:
        self._remove_from_red_list(user_id)
```
(or similar, with more comments and awaits)

Reply:
Good catch!